gobblin-data-management/src/main/java/gobblin/data/management/copy/WatermarkBasedCopyableFileFilter.java

```diff
@@ -0,0 +1,186 @@
+    this.callablePrevWatermarkMap = new Callable<Map<String, IncludeExcludeWatermark>>() {
+      @Override
+      public Map<String, IncludeExcludeWatermark> call()
+          throws Exception {
+        return getPreviousWatermarkForFilterIfNecessary(filterByWatermark, sourcestate, copyEntityWatermarkGeneratorOptional);
+      }
+    };
+  }
+
+  @Override
+  public Collection<CopyableFile> filter(FileSystem sourceFs, FileSystem targetFs,
+      Collection<CopyableFile> copyableFiles) {
+    if (!this.filterByWatermark) {
+      return copyableFiles;
+    }
+    List<CopyableFile> filteredCopyableFiles = new ArrayList<>();
+    try {
+      for (CopyableFile copyableFile : copyableFiles) {
+        Optional<WatermarkInterval> watermarkIntervalOptional =
+            CopyEntityWatermarkHelper.getCopyEntityWatermark(copyableFile, this.copyEntityWatermarkGeneratorOptional);
+        Map<String, IncludeExcludeWatermark> prevWatermarkMap =
+            STATE_TO_PREV_WATERMARK_MAP_CAHCE.get(this.sourcestate, this.callablePrevWatermarkMap);
+        String datasetURN = copyableFile.getDatasetAndPartition(this.copyableDatasetMetadata).toString();
+        if (!shouldSkipDueToWatermark(prevWatermarkMap.get(datasetURN), watermarkIntervalOptional)) {
+          log.info("Adding one copyableFile after checking watermark: " + copyableFile);
+          filteredCopyableFiles.add(copyableFile);
+        }
+      }
+      return filteredCopyableFiles;
+    } catch (IOException | ExecutionException e) {
+      throw new RuntimeException("Failed to filter copyableFiles.", e);
+    }
+  }
+
+  private boolean shouldSkipDueToWatermark(IncludeExcludeWatermark previousWatermark,
+      Optional<WatermarkInterval> curWatermarkIntervalOptional) {
+    if (curWatermarkIntervalOptional.isPresent() && previousWatermark != null) {
+      if (previousWatermark.isInclude()) {
+        return ((ComparableWatermark) (curWatermarkIntervalOptional.get().getExpectedHighWatermark()))
+            .compareTo(previousWatermark.getWatermark()) < 0;
+      } else {
+        return ((ComparableWatermark) (curWatermarkIntervalOptional.get().getExpectedHighWatermark()))
+            .compareTo(previousWatermark.getWatermark()) <= 0;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Return the mapping from datasetUrn to {@link IncludeExcludeWatermark} based on previous workunitstates if filterByWatermark is enabled and watermarkGenerator is provided.
+   * If all {@link WorkUnitState}s are committed, watermark value will be the highest high watermark among all, and this value should be excluded for next time.
+   * Otherwise, it will be the lowest low watermark among all non-committed {@link WorkUnitState}s, and this value should included for next time.
+   */
+  @VisibleForTesting
+  protected Map<String, IncludeExcludeWatermark> getPreviousWatermarkForFilterIfNecessary(boolean filterByWatermark,
```

Comment:
It is actually only instantiated once. `STATE_TO_PREV_WATERMARK_MAP_CAHCE` stores the mapping from `SourceState` reference to `Map`.